### PR TITLE
feat: new org unit dimension (TECH-707)

### PR DESCRIPTION
**Pull request by martinkrulltott**
Implements [TECH-707](https://jira.dhis2.org/browse/TECH-707)

**Relates to https://github.com/dhis2/data-visualizer-app/pull/1939**

---

### Key features

1. Implements the new org unit tree from `ui`

---

### Description

Replaces the old `d2` based org unit tree with the org unit tree plus `Checkbox`es and `MultiSelect`s from `ui`.
New props are passed in (see _breaking changes_ below), which are more in line with the other dimensions (data, period, dynamic dimensions).

Removed the unit test for org unit dimension, as it didn't really do anything and these dimensions are better tested through Cypress tests in the respective apps.

Note: `organisationUnits-dataEngine.js` renamed to `organisationUnits.js` and the old file was removed (was only used by the old org unit tree and wasn't exported externally anymore either)

---

### BREAKING CHANGE

changed props for org the unit dimension
_removed_: `current`, `d2`, `displayNameProperty`, `ouItems`, `onDeselect`
_added_: `root`, `selected`
_unaffected_: `onSelect`

---

### TODO

- [ ] Remove `apiFetchOrganisationUnits`?
- [ ] Fetch `root` from backend settings instead of passing it in as a prop?

---

### Known issues

- [x] `Failed prop type: Invalid org unit path 'ImspTQPwCqd' supplied to 'OrganisationUnitTree.initiallyExpanded[0]'` is thrown on first open. Need to discuss how to handle multiple roots first.

---

### Screenshots

image

### CODE REVIEW

File: src/api/organisationUnits.js
Comment (HendrikThePendric) [inline src/api/organisationUnits.js]: It could be worthwhile to also default to an empty object so the fn can be called without any args.
```suggestion
    params: ({ displayNameProp = 'displayName' } = {}) => ({
```

File: src/api/organisationUnits.js
Comment (HendrikThePendric) [inline src/api/organisationUnits.js]: Same as above:
```suggestion
    params: ({ displayNameProp = 'displayName' } = {}) => ({
```

File: src/api/organisationUnits.js
Comment (HendrikThePendric) [inline src/api/organisationUnits.js:32]: This `level: 1,` was already there so it must be correct, but I don't understand how having a hardcoded `level` is correct if this is a generic way to get a list of org-units.

**Review by HendrikThePendric — APPROVED** [review]
Left two minor suggestions and a question, but LGTM.

**Review by edoardo — COMMENTED** [review]
In general I would like to reach a conclusion on which way we should use `dataEngine`, `dataQuery`, perhaps that can be addressed in a separate PR for this component.

File: src/api/organisationUnits.js
Comment (martinkrulltott) [reply src/api/organisationUnits.js:32]: I think this fetches the "highest level" for the user (which is dynamic, could be one level for you and another for me, depending on our root org units). Not entirely sure, but I believe @janhenrikoverland knows more about how the backend treats this.

**Comment by martinkrulltott** [comment]
No rush to merge this and as its a breaking change its worth figuring out everything before merging. So we can hold off until we've got @edoardo's concerns answered if needed.

File: src/components/PeriodDimension/FixedPeriodSelect.js
Comment (edoardo) [inline src/components/PeriodDimension/FixedPeriodSelect.js]: Typo?
```suggestion
import PropTypes from 'prop-types'
```

and also rename in `propTypes` below.

File: src/components/TransferOption.js
Comment (edoardo) [inline src/components/TransferOption.js]:
```suggestion
import PropTypes from 'prop-types'
```

and also rename in `propTypes` below.

File: src/components/OrgUnitDimension/OrgUnitDimension.js
Comment (edoardo) [inline src/components/OrgUnitDimension/OrgUnitDimension.js]: I have a similar thing for the sharing summary.
I don't know what is the best solution, as for some languages this "hardcoded" comma-separated list of text can be a problem.
I don't have a better suggestion tho...

**Review by edoardo — APPROVED** [review]
The `dataEngine` usage can be addressed in a separate PR once we have a decision.

File: src/components/OrgUnitDimension/OrgUnitDimension.js
Comment (martinkrulltott) [reply src/components/OrgUnitDimension/OrgUnitDimension.js]: I had a long chat with Joe about this, because it started as "A," + "B," + "C"+ " selected, which is bad because they're all part of the same sentence. However with the "Selected: " + "parts", at least the sentence is split up. I could do the same as you do though, inject the comma list with interpolation instead of just concating the string.

File: src/__demo__/OrgUnitDimension.stories.js
Comment (HendrikThePendric) [inline src/__demo__/OrgUnitDimension.stories.js:26]: The StoryBook `storiesOf` API has not been deprecated, and there are no concrete plans to deprecate it in the near future ([source](https://github.com/storybookjs/storybook/blob/master/lib/core/docs/storiesOf.md)). However StoryBook recommends writing stories in the [Component Story Format (CSF)](https://storybook.js.org/docs/react/api/csf).

Personally I find CFS easier to write and there are some additional advantages in case you want to create interactive demoes (with prop-tables). 

This isn't a critical point at all, probably more of a recommendation for future PRs.

**Review by HendrikThePendric — APPROVED** [review]
LGTM. 

_Looking at some of the code in this PR made me wonder why we don't export custom hooks instead of imperative `apiFetch` functions. But that change needs to be discussed more thoroughly and is out-of-scope for this PR._

File: src/components/OrgUnitDimension/OrgUnitDimension.js
Comment (HendrikThePendric) [reply src/components/OrgUnitDimension/OrgUnitDimension.js]: Curtesy of @mediremi :
```js
const formatList = items => {
    // Wrap Intl.ListFormat in try/catch as DHIS2 locales are not always ISO 639 compliant
    try {
        const formatter = new Intl.ListFormat(i18n.language, {
            style: 'long',
            type: 'conjunction',
        })
        return formatter.format(items)
    } catch (error) {
        return items.join(', ')
    }
}
```
([Source](https://github.com/dhis2/user-app/blob/8b323c366b9e18421856f47e1ba76a4d2ad5d3db/src/containers/UserList/OrganisationUnitFilter.js#L10-L21))

Comment (martinkrulltott) [reply src/components/OrgUnitDimension/OrgUnitDimension.js]: Looks good, added 👍  Thanks!

**Pull request merged**
